core/src/main/java/org/infinispan/configuration/cache/MemoryStorageConfiguration.java

```diff
@@ -66,12 +66,15 @@ public boolean equals(Object o) {
 
       MemoryStorageConfiguration that = (MemoryStorageConfiguration) o;
 
-      return attributes.equals(that.attributes);
+      if (!attributes.equals(that.attributes)) return false;
+      return storageType == that.storageType;
```

Comment:
Any idea why `storageType` is not an attribute like the others?

Reply:
because this is how it is rendered:
```xml
<memory>
    <off-heap size="10000000" eviction="MEMORY"/>
</memory>
```

Reply:
It looks like an element to me, with its own attributes. This is a recurring issue throughout the  configuration, many things are declared as ```Attribute``` but the way it is written to the XML is random, sometimes as an attribute, sometimes as an element, sometimes as a mixture of both...

Reply:
The original intent was because each storage type would have different parameters and this way we could have validation in the parser. To be honest now with recent changes, there is no reason it should be a separate element anymore imo.

Reply:
It was an attribute before, but I moved some things around when doing the JSON stuff to bring the serialized (XML/JSON) representation closer to the class representation, in the sense that an attribute is a value, and an element is a container of values and other elements

Reply:
Oh sorry I was getting confused by xml attribute and attribute in this config class. I didn't even realize it was broken out in https://github.com/infinispan/infinispan/commit/16467c96e7fed3d0c9928a581ddb2f90467f21ca#diff-d99d40293ec9b3f29e5a8658769c6629

Reply:
Thinking about it, it should've been broken out more, since <off-heap> has its own particular attributes, same for <object> and <binary>. I simply "promoted" storage type to the top element/configuration class, but the attributes are still flattened inside it, as they were before

Reply:
@wburns I would have loved it if off-heap had `size` and object had `count` ;)

@gustavonalle IMO it's still an attribute, it's just represented as an element name in the XML, like the cache mode is represented in the XML as the name of the cache element.

Reply:
@danberindei that may be a bit more confusing for the user, but I am fine with changing this for 11.0 if we wanted.

Reply:
@danberindei I prefer ```Attribute``` written as XML/JSON attributes :)
The JSON serializer requires this currently and potentially the XML serialization could be dropped and made generic provided the randomness of the configurations are kept at a minimum

Reply:
@gustavonalle I would also prefer XML/JSSON attributes to be `Attribute` in the programmatic config and XML/JSON elements to be builder classes, but I think the right way to do that is to make the storage type an attribute in XML/JSON, not to make it a "not a builder, not an attribute" in the programmatic configuration.

Reply:
@wburns sorry, I meant `eviction="COUNT" count="100"` and `eviction="MEMORY" size="100"`, it's not really about off-heap vs object.

`eviction-count="100"` and `eviction-memory="100"` would be even more concise, the only complication with representing them as 2 attributes is we must allow one attribute to override the other when inheriting from a template but we must raise an error if both attributes are set in the same configuration.

Reply:
@danberindei Would you want to change the config and xsd for both 10.1.x and 11.x instead of this PR?

Reply:
@danberindei btw, the storage type is currently a builder: https://github.com/infinispan/infinispan/blob/487cbc942a72c28200a2efebc0636fb7ed23d760/core/src/main/java/org/infinispan/configuration/cache/MemoryStorageConfigurationBuilder.java#L20

Reply:
That's actually the `memory()` builder.